lib/danger/danger_core/executor.rb

```diff
@@ -10,61 +13,74 @@ def run(env: nil,
             base: nil,
             head: nil,
             dangerfile_path: nil,
-            danger_id: nil)
-
-      cork ||= Cork::Board.new(silent: false,
-                              verbose: false)
-
-      # Could we find a CI source at all?
-      unless EnvironmentManager.local_ci_source(@system_env)
-        abort("Could not find the type of CI for Danger to run on.".red)
-      end
-
-      # Could we determine that the CI source is inside a PR?
-      unless EnvironmentManager.pr?(@system_env)
-        cork.puts "Not a Pull Request - skipping `danger` run".yellow
-        return
-      end
+            danger_id: nil,
+            fail_on_errors: nil)
+      # Create a silent Cork instance if cork is nil, as it's likely a test
+      cork ||= Cork::Board.new(silent: false, verbose: false)
 
-      # OK, then we can have some
-      env ||= EnvironmentManager.new(@system_env, cork)
+      # OK, we now know that Danger can run in this enviroment
+      env ||= EnvironmentManager.new(system_env, cork)
       dm ||= Dangerfile.new(env, cork)
 
+      # Setup internal state
       dm.init_plugins
-
       dm.env.fill_environment_vars
 
       begin
-        dm.env.ensure_danger_branches_are_setup
-
-        # Offer the chance for a user to specify a branch through the command line
-        ci_base = base || EnvironmentManager.danger_base_branch
-        ci_head = head || EnvironmentManager.danger_head_branch
-
-        dm.env.scm.diff_for_folder(".", from: ci_base, to: ci_head)
+        # Sets up the git environment
+        dm.setup_for_running(base_branch(base), head_branch(head))
 
         # Parse the local Dangerfile
-        dm.parse(Pathname.new(dangerfile_path))
+        dm.parse Pathname.new(dangerfile_path)
 
-        post_results(dm, danger_id)
+        # Push results to the API
+        # Pass along the details of the run to the request source
+        # to send back to the code review site.
+        dm.post_results(danger_id)
+
+        # Print results in the terminal
         dm.print_results
       ensure
+
+        # Makes sure that Danger specific git branches are cleaned
         dm.env.clean_up
       end
+
+      # By default Danger will use the status API to fail a build,
+      # allowing execution to continue, this behavior isn't always
+      # optimal for everyone.
+      exit(1) if fail_on_errors && dm.failed?
+    end
+
+    def validate!
+      validate_ci!
+      validate_pr!
+    end
+
+    private
+
+    attr_reader :system_env
```

Comment:
I uses it because this helps fail fast. `@system_enb` will return `nil`, while if you do `system_enb` will raise an `NoMethodError`. 😊